src/server/app/stac_api_collection_search.py

```diff
@@ -78,21 +78,28 @@ def check_health(self) -> str:
         except APIError:
             return "cannot be opened by pystac_client"
 
-    def get_collection_metadata(self) -> Iterable[CollectionMetadata]:
-        catalog = Client.open(self.base_url)
+    def get_collection_metadata(
+        self,
+    ) -> Iterable[Union[CollectionMetadata, FederatedSearchError]]:
+        try:
+            catalog = Client.open(self.base_url)
 
-        # add /collections conformance class just in case it's missing...
-        # https://github.com/stac-utils/pystac-client/issues/320
-        # cmr-stac is still missing this conformance class
-        # https://github.com/nasa/cmr-stac/issues/236
-        # this makes it possible to iterate through all collections
-        catalog.add_conforms_to("COLLECTIONS")
+            # add /collections conformance class just in case it's missing...
+            # https://github.com/stac-utils/pystac-client/issues/320
+            # cmr-stac is still missing this conformance class
+            # https://github.com/nasa/cmr-stac/issues/236
+            # this makes it possible to iterate through all collections
+            catalog.add_conforms_to("COLLECTIONS")
 
-        return (
-            self.collection_metadata(collection)
-            for collection in catalog.get_collections()
-            if self.overlaps(collection)
-        )
+            return (
+                self.collection_metadata(collection)
+                for collection in catalog.get_collections()
+                if self.overlaps(collection)
+            )
+        except APIError as e:
+            return [
+                FederatedSearchError(catalog_url=self.base_url, error_message=str(e))
+            ]
```

Comment:
This might catch an error raised from `Client.open` or `catalog.get_collections`, but because `get_collections` returns a lazy iterator (generator), if the call doesn't raise, no matter what you attempt to catch here, you will not catch an exception raised during iteration over the returned generator, which you iterate over in a different place in the code.

Instead, consider something along these lines:

```python
        try:
            catalog = Client.open(self.base_url)

            # add /collections conformance class just in case it's missing...
            # https://github.com/stac-utils/pystac-client/issues/320
            # cmr-stac is still missing this conformance class
            # https://github.com/nasa/cmr-stac/issues/236
            # this makes it possible to iterate through all collections
            catalog.add_conforms_to("COLLECTIONS")

            for collection in catalog.get_collections():
                if self.overlaps(collection):
                    yield self.collection_metadata(collection)
        except APIError as e:
            yield FederatedSearchError(catalog_url=self.base_url, error_message=str(e))
```

Reply:
Yeah, this is exactly the kind of case I wasn't sure about. Thanks for the suggestion!